File: CPP/plugins/vehicle_base/src/create_vehicle.cpp
/**
 * @file create_vehicle.cpp
 *
 * @brief Creation of derived objects through the available plugin
 *
 * @author Antonio Mauro Galiano
 * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
 *
 */


#include <pluginlib/class_loader.hpp>
#include "vehicle_base/regular_vehicle.hpp"
#include <iostream>

int main(int argc, char** argv)
{
  // No parameters passed to the main (used to do initializing the node)
  // these two lines avoid compile warnings
  (void) argc; (void) argv;

  pluginlib::ClassLoader<vehicle_base::RegularVehicle> VehiclePluginLoader(
  "vehicle_base", "vehicle_base::RegularVehicle");

 try
  {
    // the istance of the derived class Motorbike is created through the plugin as shared ptr
    // and it's called the initialize method setting the vehicle type to "motorbike"
    std::shared_ptr<vehicle_base::RegularVehicle> motorbike = VehiclePluginLoader.createUniqueInstance(
        "vehicle_plugins::Motorbike");
    motorbike->initialize("motorbike");

    // same as motrobike but with bicycle
    std::shared_ptr<vehicle_base::RegularVehicle> bicycle = VehiclePluginLoader.createUniqueInstance(
        "vehicle_plugins::Bicycle");
    bicycle->initialize("bicycle");

    // use of the method getVehicleType through the shared ptrs of the instances derived
    // from the base class using the plugin
    std::cout << "First vehicle type is -> " << motorbike->getVehicleType() << std::endl;
    std::cout << "Second vehicle type is -> " << bicycle->getVehicleType() << std::endl;
  }
  catch(pluginlib::PluginlibException& ex)
  {
    std::cout << "The plugin failed to load for some reason. Error: " << ex.what() << std::endl;
  }

   try
  {
    // here is trying to initialize a shared ptr of a derived class which is not available with
    // the created plugin, so it gives an error
    std::shared_ptr<vehicle_base::RegularVehicle> rocket = VehiclePluginLoader.createUniqueInstance("vehicle_plugins::Rocket");
    rocket->initialize("rocket");

    std::cout << "Third vehicle type is -> " << rocket->getVehicleType() << std::endl;
  }
  catch(pluginlib::PluginlibException& ex)
  {
    std::cout << "The plugin failed to load for some reason. Error: " << ex.what() << std::endl;
  }

  return 0;
}

File: PYTHON/plugins/vehicle_plugins_py/vehicle_plugins_py/create_vehicle.py
#  *
#  * @file create_vehicle.py
#  *
#  * @brief Creation of derived objects through the available plugins
#  *
#  * @author Antonio Mauro Galiano
#  * Contact: https://www.linkedin.com/in/antoniomaurogaliano/
#  *
#

import sys
try:
    from pkg_resources import iter_entry_points
except ImportError:
    # For newer Python versions, use importlib.metadata
    try:
        from importlib.metadata import entry_points
        def iter_entry_points(group):
            return entry_points(group=group)
    except ImportError:
        # Fallback for older Python
        from importlib_metadata import entry_points
        def iter_entry_points(group):
            return entry_points(group=group)

from vehicle_base_py import RegularVehicle


def load_plugin(plugin_name: str) -> RegularVehicle:
    """
    Load a plugin by name using entry points.
    
    Args:
        plugin_name: Name of the plugin (e.g., 'motorbike', 'bicycle', 'truck')
    
    Returns:
        An instance of RegularVehicle
    
    Raises:
        ValueError: If the plugin is not found
    """
    # Load plugin using entry points
    for entry_point in iter_entry_points('vehicle_base_py.plugins', plugin_name):
        factory = entry_point.load()
        return factory()
    raise ValueError(f"Plugin '{plugin_name}' not found")


def main():
    """Main function to demonstrate plugin usage."""
    print("=== Loading Available Plugins ===")
    try:
        # Create instances of derived classes through the plugin system
        # The instance of the derived class Motorbike is created through the plugin
        # and it's called the initialize method setting the vehicle type to "motorbike"
        motorbike = load_plugin('motorbike')
        motorbike.initialize("motorbike")

        # Same as motorbike but with bicycle
        bicycle = load_plugin('bicycle')
        bicycle.initialize("bicycle")

        # Use of the method get_vehicle_type through the instances derived
        # from the base class using the plugin
        print(f"First vehicle type is -> {motorbike.get_vehicle_type()}")
        print(f"Second vehicle type is -> {bicycle.get_vehicle_type()}")
        print("✓ Successfully loaded Motorbike and Bicycle plugins")
    except Exception as ex:
        print(f"ERROR: The plugin failed to load. Error: {ex}")
        return 1

    print("\n=== Demonstrating Error Handling (Intentional) ===")
    print("Attempting to load non-existent 'Rocket' plugin...")
    try:
        # Here is trying to initialize a derived class which is not available with
        # the created plugin, so it gives an error - this is intentional to demonstrate error handling
        rocket = load_plugin('rocket')
        rocket.initialize("rocket")
        print(f"Third vehicle type is -> {rocket.get_vehicle_type()}")
    except Exception as ex:
        print(f"Expected error caught: {ex}")
        print("✓ Error handling works correctly - plugin system properly rejects non-existent plugins")

    return 0


if __name__ == '__main__':
    sys.exit(main())

